### PR TITLE
Add settings module support for PlatformIO

### PR DESCRIPTION
**Pull request by YegorBobrovski**
discussed [on forum](https://forum.simbaos.org/showthread.php?tid=17)

usage:
1. add valid *.ini file with settings
2. set relative or absolute path to this file in board's options, for example:
```
[env:esp12e]
settings_path = ../settings/12e.ini

[env:esp8285]
settings_path = ~/settings/8285.ini
```

### CODE REVIEW

File: bin/simbagen.py [hunk 1]
```diff
@@ -168,11 +168,13 @@ def __init__(self, filename, endianess):
...
 
-        if filename is None:
+        if not filename:
             return
 
         settings_parser = ConfigParser()
-        settings_parser.read(filename)
+        if not settings_parser.read(filename):
+            print("Failed to load config file from {}".format(filename))
+            return
```
Comment (eerimoq) [inline bin/simbagen.py]: Nice catch, but should we just call `sys.exit("Failed to load config file '{}'.".format(filename))` instead of the print and return?

If the settings file parsing failed this script should fail.

File: bin/simbagen.py [hunk 2]
```diff
@@ -168,11 +168,13 @@ def __init__(self, filename, endianess):
         self.endianess = endianess
         self.settings = OrderedDict()
 
-        if filename is None:
+        if not filename:
```
Comment (eerimoq) [inline bin/simbagen.py]: If `--settings` is given to the script it should be a valid filename. Giving an empty string should not be treated as a filename (even if the file cannot exist).

We should undo this change and make the platformio build script only give `--settings` if a settings file is configured in `platformio.ini`.

File: make/platformio.sconscript [hunk 3]
```diff
@@ -4260,8 +4260,16 @@ source_files = []
 for src in env.LookupSources(variant_dir, src_dir, True, src_filter):
     source_files.append(env.Object(src))
 
+# get settings file path
+settings_ini_path = ARGUMENTS.get('SETTINGS_PATH', "").decode('base64')
```
Comment (eerimoq) [inline make/platformio.sconscript]: Rename `settings_ini_path` and `settings_path` to `settings_ini`, as that is how the make variable is called in the make build system, just to make the naming consistent throughout the code base. `settings_path` may be a better name let's not change it to that now.

File: make/platformio.sconscript [hunk 4]
```diff
@@ -4260,8 +4260,16 @@ source_files = []
...
+if settings_ini_path:
+    settings_ini_file = env.File(settings_ini_path, env['PROJECT_DIR'])
+    settings_ini_path = str(settings_ini_file)
```
Comment (eerimoq) [inline make/platformio.sconscript]: Merge the two lines into one by calling `str()` on the file object on the same line.

File: make/platformio.sconscript [hunk 5]
```diff
@@ -4260,8 +4260,16 @@ source_files = []
...
+    if not exists(settings_ini_path):
+        settings_ini_path = ""
```
Comment (eerimoq) [inline make/platformio.sconscript]: The build should fail if a non-existing settings file is given.

File: make/platformio.sconscript [hunk 6]
```diff
@@ -4271,9 +4279,13 @@ env.Command(SIMBA_GEN_C,
              '--board "$BOARD_DESC" '
              '--mcu "$MCU_DESC" '
              '--endianess little '
-             '--eeprom-soft-chunk-size 2048 '))
+             '--eeprom-soft-chunk-size 2048 '
+             '--settings "{}" '.format(settings_ini_path)))
```
Comment (eerimoq) [inline make/platformio.sconscript]: See comment in simbagen.py.

File: make/platformio/platformio.py [hunk 7]
```diff
@@ -407,20 +408,36 @@ def setup_board_nano32(env):
...
              '--name "$NAME" '
              '--version "$VERSION" '
              '--board "$BOARD_DESC" '
              '--mcu "$MCU_DESC" '
              '--endianess little '
-             '--eeprom-soft-chunk-size 2048 '))
+             '--eeprom-soft-chunk-size 2048 ')
+
+# get settings file path
+settings_ini = ARGUMENTS.get('SETTINGS_PATH')
```
Comment (eerimoq) [inline make/platformio/platformio.py]: Hmm, shall we rename this to ``SETTING_INI`` as well to make it consistent?

File: make/platformio/platformio.py [hunk 8]
```diff
@@ -407,20 +408,36 @@ def setup_board_nano32(env):
...
+if settings_ini is not None:
+    try:
+        settings_ini = str(env.File(settings_ini.decode('base64'), 
+                                    env['PROJECT_DIR']))
+    except Exception, e:
+        sys.exit("Failed to load config file. " + str(e))
+    gen_command_action += '--settings "{}" '.format(str(settings_ini))
```
Comment (eerimoq) [inline make/platformio/platformio.py]: Since this is a format string (``PLATFORMIO_SCONSSCRIPT_FMT``) all ``{`` and ``}`` have to be escaped. Change ``{}`` to ``{{}}``.

File: make/platformio/platformio.py (hunk 7 above)
Comment (YegorBobrovski) [inline make/platformio/platformio.py]: Sorry, misunderstood your previous review. I thought it should be named SETTINGS_PATH since now.

Ok, I'll change it.

Comment (eerimoq) [inline make/platformio/platformio.py]: Ahh okay, thanks! =)

File: make/platformio/platformio.py [hunk 9]
```diff
@@ -407,20 +408,36 @@ def setup_board_nano32(env):
...
-             '--eeprom-soft-chunk-size 2048 '))
+             '--eeprom-soft-chunk-size 2048 ')
+
+# get settings file path
+settings_ini = ARGUMENTS.get('SETTINGS_INI')
+if settings_ini is not None:
+    try:
+        settings_ini = str(env.File(settings_ini.decode('base64'), 
+                                    env['PROJECT_DIR']))
+    except Exception, e:
```
Comment (eerimoq) [inline make/platformio/platformio.py]: Hmm, ``Exception, e`` is the old Python syntax, the new is ``Exception as e``. I guess it's good to use the new syntax.
